modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/ReadWriteTransactionImpl.java

```diff
@@ -88,8 +93,12 @@ public IgniteBiTuple<ClusterNode, Long> enlist(ReplicationGroupId replicationGro
     /** {@inheritDoc} */
     @Override
     protected CompletableFuture<Void> finish(boolean commit) {
+        if (!finishFut.compareAndSet(null, new CompletableFuture<>())) {
+            return finishFut.get();
+        }
+
         // TODO: https://issues.apache.org/jira/browse/IGNITE-17688 Add proper exception handling.
-        return CompletableFuture
+        CompletableFuture<Void> mainFinishFut = CompletableFuture
```

Comment:
Naming is a little bit confusing. Do you have any other options? Up to you, though.

Reply:
For me it is good. And I don't want to wrack my brain :)